File: 409/LongestPalindrome.java
import java.util.HashMap;

class LongestPalindrome {
	// 23 ms
	HashMap<Character, Integer> counter;

	public int longestPalindrome(String s) {
		int ret = 0;

		counter = new HashMap<Character, Integer>();

		// count char occurences
		for (Character c: s.toCharArray()) {
			if (counter.containsKey(c))
				counter.put(c, counter.get(c)+1);
			else 
				counter.put(c, 1);
		}

		// count even and odd chars
		int n_odd = 0;
		for (Character c: counter.keySet()) {
			int f = counter.get(c);
			if (f%2 == 0)
				ret += f;
			else {
				ret += f-1;
				n_odd += 1;
			}
		}

		// find max odd chars
		int max_f = 0;
		char max_char;
		for (Character c: counter.keySet()) {
			int f = counter.get(c);
			if (f%2 == 1)
				if (f > max_f) {
					max_f = f;
					max_char = c;
				}
		}

		if (n_odd > 0)
			ret += 1;
		return ret;
		}
	}

File: 409/longest_palindrome.py
class Solution(object):
	def longestPalindrome(self, s):
		"""
		:type s: str
		:rtype: int
		"""
		counter = dict()
		for c in s:
			if c in counter:
				counter[c] += 1
			else:
				counter[c] = 1

		return self.n_even_chars(counter) + self.n_max_odd_chars(counter)

	def n_even_chars(self, counter):
		ret = 0
		for c, f in counter.iteritems():
			if f%2==0:
				ret += f

		return ret

	def n_max_odd_chars(self, counter):
		ret = 0
		max_char = ''
		for c, f in counter.iteritems():
			if f%2==1:
				if ret < f:
					ret = f
					max_char = c

		if max_char in counter:
			del counter[max_char]

		for c, f in counter.iteritems():
			if f%2==1:
				ret += f-1

		return ret


if __name__=="__main__":
	s = Solution()
	print s.longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth")
	# print s.longestPalindrome("abccccdd")